File: integrations/js/packages/core/src/estimate.ts
import { createHmac } from "node:crypto";

export interface EstimateResult {
  estimatedPrice: number;
  observables: Record<string, unknown>;
  priceSignature?: string;
}

export interface Signer {
  keyId: string;
  signEstimate(price: number, observables: Record<string, unknown>): string;
}

export class HmacSigner implements Signer {
  constructor(public readonly keyId: string, private readonly secret: Buffer) {}

  signEstimate(price: number, observables: Record<string, unknown>): string {
    const header = { alg: "HS256", kid: this.keyId, typ: "JOSE" } satisfies Record<string, unknown>;
    const payload = { price, observables } satisfies Record<string, unknown>;
    const encodedHeader = b64url(JSON.stringify(header));
    const encodedPayload = b64url(JSON.stringify(payload));
    const signingInput = `${encodedHeader}.${encodedPayload}`;
    const signature = createHmac("sha256", this.secret).update(signingInput).digest();
    return `${signingInput}.${b64url(signature)}`;
  }

  secretBytes(): Buffer {
    return Buffer.from(this.secret);
  }
}

export class JWKSManager {
  private readonly signers = new Map<string, HmacSigner>();

  register(signer: HmacSigner) {
    this.signers.set(signer.keyId, signer);
  }

  resolve(keyId: string) {
    return this.signers.get(keyId);
  }

  jwks() {
    return {
      keys: [...this.signers.values()].map((signer) => ({
        kid: signer.keyId,
        kty: "oct",
        alg: "HS256",
      })),
    };
  }
}

export function estimate(options: {
  estimatedPrice: number;
  observables?: Record<string, unknown>;
  signer?: Signer;
}): EstimateResult {
  const observables = options.observables ?? {};
  const priceSignature = options.signer?.signEstimate(options.estimatedPrice, observables);
  return {
    estimatedPrice: round(options.estimatedPrice),
    observables,
    priceSignature,
  } satisfies EstimateResult;
}

export function verifySignature(
  token: string,
  resolver: (kid: string) => Buffer | undefined,
): boolean {
  const parts = token.split(".");
  if (parts.length !== 3) return false;
  const [encodedHeader, encodedPayload, encodedSignature] = parts;
  let header: { kid?: string; alg?: string };
  try {
    header = JSON.parse(Buffer.from(padB64(encodedHeader), "base64url").toString("utf8"));
  } catch {
    return false;
  }
  if (!header.kid) return false;
  const secret = resolver(header.kid);
  if (!secret) return false;
  const signingInput = `${encodedHeader}.${encodedPayload}`;
  const expected = createHmac("sha256", secret).update(signingInput).digest();
  const provided = Buffer.from(padB64(encodedSignature), "base64url");
  return timingSafeEqual(expected, provided);
}

function round(value: number): number {
  return Math.round(value * 1_000_000) / 1_000_000;
}

function b64url(input: string | Buffer): string {
  const buffer = typeof input === "string" ? Buffer.from(input) : input;
  return buffer.toString("base64").replace(/=+$/g, "").replace(/\+/g, "-").replace(/\//g, "_");
}

function padB64(input: string): string {
  const pad = (4 - (input.length % 4)) % 4;
  return input + "=".repeat(pad);
}

function timingSafeEqual(a: Buffer, b: Buffer): boolean {
  if (a.length !== b.length) return false;
  let result = 0;
  for (let idx = 0; idx < a.length; idx += 1) {
    result |= a[idx] ^ b[idx];
  }
  return result === 0;
}

File: integrations/python/tribute_core/estimate.py
"""Estimate pipeline helpers: price construction, signing, and verification."""

from __future__ import annotations

import base64
import hmac
import json
from dataclasses import dataclass
from decimal import Decimal, ROUND_HALF_UP
from hashlib import sha256
from typing import Any, Callable, Dict, Mapping, Optional, Protocol


def _b64url(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode("ascii")


@dataclass
class EstimateResult:
    """Normalized estimate payload returned to integrators."""

    estimated_price: Decimal
    observables: Mapping[str, Any]
    price_signature: Optional[str]

    def to_dict(self) -> Dict[str, Any]:
        payload = {
            "estimated_price": str(self.estimated_price.quantize(Decimal("0.000001"), rounding=ROUND_HALF_UP)),
            "observables": self.observables,
        }
        if self.price_signature:
            payload["price_signature"] = self.price_signature
        return payload


class Signer(Protocol):
    key_id: str

    def sign_estimate(self, price: Decimal, observables: Mapping[str, Any]) -> str:
        ...


class HMACSigner:
    """JWS HS256 signer for price estimates (suitable for PoC usage)."""

    def __init__(self, *, key_id: str, secret: bytes):
        self.key_id = key_id
        self._secret = secret

    def sign_estimate(self, price: Decimal, observables: Mapping[str, Any]) -> str:
        header = {"alg": "HS256", "kid": self.key_id, "typ": "JOSE"}
        payload = {
            "price": str(price),
            "observables": observables,
        }
        encoded_header = _b64url(json.dumps(header, separators=(",", ":")).encode("utf-8"))
        encoded_payload = _b64url(json.dumps(payload, separators=(",", ":")).encode("utf-8"))
        signing_input = f"{encoded_header}.{encoded_payload}".encode("ascii")
        digest = hmac.new(self._secret, signing_input, sha256).digest()
        encoded_signature = _b64url(digest)
        return f"{encoded_header}.{encoded_payload}.{encoded_signature}"

    @property
    def secret(self) -> bytes:
        return self._secret


class JWKSManager:
    """In-memory JWKS manager for rotating signing keys."""

    def __init__(self):
        self._signers: Dict[str, HMACSigner] = {}

    def register(self, signer: HMACSigner) -> None:
        self._signers[signer.key_id] = signer

    def resolve(self, key_id: str) -> Optional[HMACSigner]:
        return self._signers.get(key_id)

    def jwks(self) -> Dict[str, Any]:
        return {
            "keys": [
                {
                    "kid": signer.key_id,
                    "kty": "oct",
                    "alg": "HS256",
                }
                for signer in self._signers.values()
            ]
        }


def estimate(
    *,
    estimated_price: Decimal,
    observables: Optional[Mapping[str, Any]] = None,
    signer: Optional[Signer] = None,
) -> EstimateResult:
    """Construct an estimate response and optionally sign it."""

    observables = observables or {}
    signature = signer.sign_estimate(estimated_price, observables) if signer else None
    return EstimateResult(
        estimated_price=estimated_price,
        observables=dict(observables),
        price_signature=signature,
    )


def verify_signature(
    *,
    token: str,
    key_resolver: Callable[[str], Optional[bytes]],
) -> bool:
    """Validate an HS256 signature and return True when it matches."""

    try:
        encoded_header, encoded_payload, encoded_signature = token.split(".")
    except ValueError:
        return False

    try:
        header = json.loads(base64.urlsafe_b64decode(_pad_b64(encoded_header)))
    except json.JSONDecodeError:
        return False

    kid = header.get("kid")
    if not kid:
        return False

    secret = key_resolver(str(kid))
    if secret is None:
        return False

    signing_input = f"{encoded_header}.{encoded_payload}".encode("ascii")
    expected = hmac.new(secret, signing_input, sha256).digest()
    provided = base64.urlsafe_b64decode(_pad_b64(encoded_signature))
    return hmac.compare_digest(expected, provided)


def _pad_b64(segment: str) -> bytes:
    padding = "=" * ((4 - len(segment) % 4) % 4)
    return (segment + padding).encode("ascii")
